### PR TITLE
Add btf metadata inspection

### PR DESCRIPTION
**Pull request by mendess**
Adds BTF metadata inspection to the library.

The organization is as follows

### `Btf`
This struct maps to the `libbpf_sys::btf` type, it let's you search for types in the btf data

### `BtfType`
This struct maps to `libbpf_sys::btf_type` struct and it let's you query for the data that is common among all btf types

### Extra attributes (size/type)
There is a union in `libbbpf_sys::btf_type` whose members must be accessed depending on the `BpfKind`, I encoded this in the type system through two traits, `ReferencesType` and `HasSize`.

## `src/btf/types.rs`
This file implements a different struct for each of the btf types, this also mimics the way `libbpf` is organized, since there is a different struct for each. The difference here is that every struct has a different type that derefs to the original such that it only increases the amount of information you have. For all of these there is a `TryFrom` impl that let's you convert from `BtfType` to the concrete type.

# Discussion and design decision details

Name's are always: `Option<&Cstr>`, the reasoning for each of them is:
- `Option`: I inspected a bit of the code in libbpf and it is possible for `btf__name_at_offset` to return NULL, since I didn't know if there was some guarantee that this function could never return NULL when I passed it some offset that came from btf itself I didn't risk it, but this is on `Option` I'd like to get rid of
- `CStr`: I though a bit about what to do here, using `str` forces UTF-8, but I don't know if I can assume UTF-8, since these strings are effectively user input. The 3 options I had were
    - `CStr`: (almost) 0 cost, leaves it up to the user how they want to interpret the bytes, but in this form it already implements `Display`
    - `str`: will work most of the time but take the agency out of the library user, since we will have no option but to panic if the bytes are not valid utf8
    - `Result<str, Utf8Error>`: does the check upfront for the user, but is annoying to work with out of the gate for the user
    - `String`: allocates memory, doesn't unwrap (we can use the lossy conversion), but assumes too much and is lossy

### CODE REVIEW

**Comment by danielocfb** [comment]
So we already have BTF functionality [here](https://github.com/libbpf/libbpf-rs/tree/master/libbpf-cargo/src/btf) (also see https://github.com/libbpf/libbpf-rs/issues/219). Are you aware of that? If so, in which way is it insufficient? I haven't looked at what is more logical/cleaner, but at the very least we need to converge on one, I'd say.

**Comment by mendess** [comment]
@danielocfb no, I was not aware of this since I don't use `libbpf-cargo`, I only use `libbpf-rs` in my project. I also admit I should have looked at that issue, I did find it but forgot about it when I decided to give this implementation a go.

That implementation (`libbpf-cargo`) seems to be doing a lot of work that `libbpf` is already doing (some of the parsing) whereas mine is trying to reuse as much C code as possible, I've also tried to use Rust's lifetimes to encode the invariants needed to make this kind of wrapper (as mentioned in that issue).

Would it make sense to implement btf API's here in `libbpf-rs` and refactor `libbpf-cargo` to use this new API?

> which way is it insufficient?
as far as I can tell it only allows loading from a file/raw bytes? I needed to also inspect btf data of running programs.

File: libbpf-rs/src/btf.rs [hunk 1]
```diff
@@ -0,0 +1,440 @@
+//! Parse and introspect btf information, from files or loaded objects.
```
Comment (danielocfb) [inline libbpf-rs/src/btf.rs:1]: Let's create a proper module structure below `src/btf/`? I find it confusing to have a sub-directory without `mod.rs` and don't believe it's encouraged Rust project layout. E.g.,
`src/btf.rs` -> `src/btf/mod.rs`

File: libbpf-rs/src/btf.rs [hunk 2]
```diff
@@ -0,0 +1,440 @@
...
+///   - [FuncProto](BtfKind::FuncProto),
+///   - [Var](BtfKind::Var),
+///   - [DeclTag](BtfKind::DeclTag),
+///   - [TypeTag](BtfKind::TypeTag)
+pub unsafe trait ReferencesType<'btf, 's>: Deref<Target = BtfType<'btf, 's>>
+where
+    'btf: 's,
+{
+    /// The refered type.
+    fn references_type_with_id(&self) -> TypeId {
```
Comment (danielocfb) [inline libbpf-rs/src/btf.rs:437]: Perhaps call it `referenced_type_id`?

**Review by danielocfb — COMMENTED** [review]
I haven't reviewed everything in detail, but overall I think it's a solid patch and that it would make sense to have this functionality exposed. Thanks for working on it! I left a few comments.

My biggest issue is all the macro magic, that is really hard to follow and make sense of (alas, I don't really have a better suggestion).

That being said, we most certainly should convert `libbpf-cargo` over to using this module before landing (and identify any gaps that may still be there) instead of having two implementations around. I can help with that if that's nothing you are interested in doing. I'd also strongly prefer were have actual tests for this logic (though integration with `libbpf-cargo` could potentially suffice).
Lastly, can you please split this pull request up more (and rebase against `master` and fix CI) to make it easier digestible? E.g., first `Btf` type, then `BtfType` type, then individual concrete types, along with required macros, etc.

**Comment by danielocfb** [comment]
> Name's are always: `Option<&Cstr>`, the reasoning for each of them is:
> 
> * `Option`: I inspected a bit of the code in libbpf and it is possible for `btf__name_at_offset` to return NULL, since I didn't know if there was some guarantee that this function could never return NULL when I passed it some offset that came from btf itself I didn't risk it, but this is on `Option` I'd like to get rid of

That's for `Btf::name_at`, correct? Some types logically don't have names from what I understand, e.g., pointers. They apparently still map to an empty string in the string table, so you wouldn't get back a NULL pointer. So I'd think that conceptually an `Option` is fine and we may want to consider mapping `""` to `None`. We could go so far to model this as a trait like you did for size/referenced types, but that may be too much and potentially inflexible/confusing in the presence of types that may or may not have a name (anonymous `struct`s).

**Comment by danielocfb** [comment]
> `CStr`: I though a bit about what to do here, using `str` forces UTF-8, but I don't know if I can assume UTF-8, since these strings are effectively user input. The 3 options I had were
> * `CStr`: (almost) 0 cost, leaves it up to the user how they want to interpret the bytes, but in this form it already implements `Display`
> * `str`: will work most of the time but take the agency out of the library user, since we will have no option but to panic if the bytes are not valid utf8
> * `Result<str, Utf8Error>`: does the check upfront for the user, but is annoying to work with out of the gate for the user
> * `String`: allocates memory, doesn't unwrap (we can use the lossy conversion), but assumes too much and is lossy

My take is: `CStr` seems fine to me. We could use `Cow<'_, str>` which would be a reasonable middle ground (I don't think we would ever encounter non-UTF-8, but at least we wouldn't panic if we were), but given how [easy](https://doc.rust-lang.org/std/ffi/struct.CStr.html#method.to_string_lossy) the conversion is, that could probably be done be users equally well, if need be.

File: libbpf-rs/src/btf.rs [hunk 3]
```diff
@@ -0,0 +1,440 @@
...
+            libbpf_sys::bpf_obj_get_info_by_fd(
+                fd.as_raw_fd(),
+                &mut info as *mut _ as *mut _,
+                &mut (size_of::<libbpf_sys::bpf_prog_info>() as u32) as *mut _,
+            )
+        };
+        if err < 0 {
+            return Err(Error::System(-err));
+        }
+        nix::unistd::close(fd).map_err(|e| Error::System(e as _))?;
```
[the remark this answers is not in the document]
Comment (mendess) [reply libbpf-rs/src/btf.rs:145]: I wanted to, and actually pushed a version with it, but I hit the MSRV, I also like the OwnedFd/BorrowedFd and I've been using it in my project. If you think the bump to 1.63 is okay I'll happily do it

**Comment by mendess** [comment]
thanks for this thorough review! I agree with 99% of what you pointed out and sorry for all my typos :sweat_smile: 
I will make all of these changes and I'm going to try to find the time to refactor libbpf-cargo to use this.

File: libbpf-rs/src/btf.rs (hunk 1 above)
Comment (mendess) [reply libbpf-rs/src/btf.rs:1]: this layout was actually introduced as a replacement for the `mod.rs` layout. I personally don't have much strong feelings about either one, iirc the reason for this layout is that it's better for people using fuzzy finders in their IDE (typing mod.rs wasn't usefull because it showed every mod) but I use a fuzzy finder and I can work the both layouts so it never really bothered me. If you prefer `src/btf/mod.rs` I can change it

File: libbpf-rs/src/btf.rs [hunk 4]
```diff
@@ -0,0 +1,440 @@
...
+///   - [Ptr](BtfKind::Ptr)
+///   - [Typedef](BtfKind::Typedef),
+///   - [Volatile](BtfKind::Volatile),
+///   - [Const](BtfKind::Const),
+///   - [Restrict](BtfKind::Restrict),
+///   - [Func](BtfKind::Func),
+///   - [FuncProto](BtfKind::FuncProto),
+///   - [Var](BtfKind::Var),
+///   - [DeclTag](BtfKind::DeclTag),
+///   - [TypeTag](BtfKind::TypeTag)
```
[the remark this answers is not in the document]
Comment (mendess) [reply libbpf-rs/src/btf.rs:431]: Makes sense. I though about making this a sealed trait so users can't implement the trait anyway. The problem is that I couldn't find any docs explaining which type's have which attribute set in the union, all I found was the source code of the [header](https://github.com/libbpf/libbpf/blob/b5e9722ec268702b2ec62863222277148e4fe656/include/uapi/linux/btf.h#L42-L52). which are technically docs. Do you think I should link this here?

File: libbpf-rs/src/btf.rs [hunk 5]
```diff
@@ -0,0 +1,440 @@
...
+
+/// A undiscriminated btf type
+///
+/// For the documentation of the lifetime `'btf` see [`Btf`].
+///
+/// The lifetime `'s` guards against this type outliving the [`Btf`] instance it came from.
+#[derive(Clone, Copy)]
+pub struct BtfType<'btf, 's>
+where
+    'btf: 's,
```
[the remark this answers is not in the document]
Comment (mendess) [reply libbpf-rs/src/btf.rs:300]: You're right, I was trying to be smart here because the lifetime of the name is sometimes longer than the lifetime of the Btf, in the case where the name comes from a `bpf_object`, but sadly without going full on type magic, I couldn't make it work (hence the comment on the `name` field).

What I failed to notice was that I could trim the lifetime after that attempt failed and indeed I can make this simpler.

File: libbpf-rs/src/btf.rs (hunk 3 above)
[the remark this answers is not in the document]
Comment (danielocfb) [reply libbpf-rs/src/btf.rs]: Yeah, I think it's fine. Let's do it as a separate commit and add a `CHANGELOG.md` entry and it should be no problem. I think 4c7034e919f74ba927af0504a23770beff75b545 should illustrate everything that needs to be touched.

File: libbpf-rs/src/btf.rs [hunk 6]
```diff
@@ -0,0 +1,440 @@
...
+
+/// The id of a btf type.
+#[derive(Debug, Clone, Copy, PartialEq, Eq, PartialOrd, Ord, Hash)]
+pub struct TypeId(u32);
+
+impl From<u32> for TypeId {
+    fn from(s: u32) -> Self {
+        Self(s)
+    }
+}
```
[the remark this answers is not in the document]
Comment (danielocfb) [reply libbpf-rs/src/btf.rs:97]: Sounds good.

File: libbpf-rs/src/btf.rs (hunk 4 above)
[the remark this answers is not in the document]
Comment (danielocfb) [reply libbpf-rs/src/btf.rs]: > I though about making this a sealed trait so users can't implement the trait anyway.

I think sealing the trait is a good suggestion, considering that it's not meant to be a user-extensible format.

> Do you think I should link this here?

Linking some reference is a great idea, as I also always struggle finding relevant information. I would htink that https://www.kernel.org/doc/html/latest/bpf/btf.html is the authoritative user (developer) facing documentation and would hope that it's reasonably stable to even link to individual sections.

**Comment by danielocfb** [comment]
> thanks for this thorough review! I agree with 99% of what you pointed out and sorry for all my typos sweat_smile I will make all of these changes and I'm going to try to find the time to refactor libbpf-cargo to use this.

Sounds good. Thanks again! Please let me know when you think it's ready and I'll give it another pass.

File: libbpf-rs/src/btf.rs (hunk 1 above)
Comment (mendess) [reply libbpf-rs/src/btf.rs]: mine does too, best fuzzy is just `find | fzf` or equivalent xD I'm gonna go the mod.rs route. I've personally tried giving the non `mod.rs` route but I think it looks more confusing when you're looking at the dir structure

File: libbpf-rs/src/btf.rs (hunk 3 above)
[the remark this answers is not in the document]
Comment (mendess) [reply libbpf-rs/src/btf.rs]: I tried using cast more in the code, but this [particular method](https://doc.rust-lang.org/std/primitive.pointer.html#method.cast_mut) is only available from 1.65 up, do I raise the MSRV that high?

[the remark this answers is not in the document]
Comment (danielocfb) [reply libbpf-rs/src/btf.rs]: > I tried using cast more in the code, but this [particular method](https://doc.rust-lang.org/std/primitive.pointer.html#method.cast_mut) is only available from 1.65 up, do I raise the MSRV that high?

May not be worth it right now. I'd say let's stick to `1.63` until we have more formally defined the MSRV policy.

**Comment by mendess** [comment]
@danielocfb I think it's done. Can you give it another look?

I hope I split the commits well :)

**Review by danielocfb — COMMENTED** [review]
I think your changes look great! I left a few more minor comments that we probably should address.

Other than that the only other question is what are your plans for testing and the `libbpf-cargo` integration as mentioned beforehand?

**Comment by mendess** [comment]
> I think your changes look great! I left a few more minor comments that we probably should address.
> 
> Other than that the only other question is what are your plans for testing and the `libbpf-cargo` integration as mentioned beforehand?

Before addressing these changes I'm going to push a few more changes that to this PR stemming from the refactoring of `libbpf-cargo`. I needed some more features in this implementation, as well as, some things I got wrong from the docs.

I'm currently trying to make the examples and tests pass. I think we can consider `libbpf-cargo` pretty good integration testing from what I've seen, it exercises a lot of the API. If you think I should introduce more tests I'm willing to do it. Can you maybe provide some guidance on the best way to tests this? Since I'm going to need some btf programs/objects to test on.

**Comment by danielocfb** [comment]
> I'm currently trying to make the examples and tests pass. I think we can consider `libbpf-cargo` pretty good integration testing from what I've seen, it exercises a lot of the API.

Yeah, I agree.

> If you think I should introduce more tests I'm willing to do it. Can you maybe provide some guidance on the best way to tests this? Since I'm going to need some btf programs/objects to test on.

I don't think you have to, it's not a blocker for the merge. We probably should eventually, but we can open an issue for the addition.

I think the way we would test the bindings is by using `libbpf` APIs to create "raw" BTF. Here is how they can be used: https://github.com/torvalds/linux/blob/cb7f5b41f8341148050fe63e27cf52aa4f1519ad/tools/testing/selftests/bpf/prog_tests/btf_write.c#L7-L353

There is no need to write polished bindings for these and expose them publicly at this point, of course. They would just be used for testing purposes.

File: libbpf-rs/src/btf/mod.rs [hunk 7]
```diff
@@ -0,0 +1,149 @@
...
+
+impl Btf<'static> {
+    /// Load the btf information from an ELF file.
+    pub fn from_path<P: AsRef<Path>>(path: P) -> Result<Self> {
+        let path = {
+            let mut v = path.as_ref().as_os_str().as_bytes().to_vec();
+            v.push(0);
+            CString::from_vec_with_nul(v).map_err(|_| {
+                Error::InvalidInput(format!("invalid path {:?}, has null bytes", path.as_ref()))
+            })?
```
[the remark this answers is not in the document]
Comment (mendess) [reply libbpf-rs/src/btf/mod.rs:132]: Although this path to convert it is really roundabout, it's the only way that doesn't go through an utf8 check. Which technically makes this capable of accepting more paths.

File: libbpf-cargo/Cargo.toml [hunk 8]
```diff
@@ -35,7 +35,7 @@ cargo_metadata = "0.15.0"
 libbpf-sys = { version = "1.0.3" }
 libbpf-rs = { version = "0.20", path = "../libbpf-rs" }
 memmap2 = "0.5"
-num_enum = "0.5"
+num_enum = "0.5.1"
```
[the remark this answers is not in the document]
Comment (mendess) [reply libbpf-cargo/Cargo.toml:38]: I'm going to actually drop this commit, as I no longer need to bump it.

File: libbpf-rs/src/btf/mod.rs [hunk 9]
```diff
@@ -0,0 +1,149 @@
...
+        };
+        let ptr = create_bpf_entity_checked(|| unsafe {
+            libbpf_sys::btf__parse_elf(path.as_ptr(), std::ptr::null_mut())
```
[the remark this answers is not in the document]
Comment (mendess) [reply libbpf-rs/src/btf/mod.rs:135]: from [this](https://github.com/libbpf/libbpf/blob/f8cd00f61302ca4d8645e007acf7f546dc323b33/src/btf.c#L1048-L1056) code in libbpf  (and reading the code of the function they delegate to) I concluded it was okay to pass null to that parameter. Since that function lacked documentation I chose to pass null and not provide a way for the caller to override that parameter for now, since I didn't understand it. In the future a `from_path_with_ext` method could be added to support passing a value there.

**Comment by mendess** [comment]
@danielocfb I've fixed everything you commented on. If the changes were simple typos then I rebased them into the original commits to reduce noise. Otherwise I made a small commit for each change. Hopefully that makes reviewing this already massive PR easier. The first new commit is [this one](https://github.com/libbpf/libbpf-rs/pull/390/commits/b18411072394f86252ec72d54fef54d297c40aa0)

I've already refactered libbpf-cargo to use this module instead. What do you think is the best way to publish those changes? Should I make a PR that points at this one?

**Comment by danielocfb** [comment]
> @danielocfb I've fixed everything you commented on. If the changes were simple typos then I rebased them into the original commits to reduce noise. Otherwise I made a small commit for each change. Hopefully that makes reviewing this already massive PR easier. The first new commit is [this one](https://github.com/libbpf/libbpf-rs/pull/390/commits/aa19b79eeda2b467f63219b57cf3abe794351979)

Thanks! Next time feel free to just merge them in (in line with suggestions in https://github.com/libbpf/libbpf-rs/blob/master/CONTRIBUTING.md#submitting-a-pull-request). I can always just diff the revisions directly for review. Will take another look soon, hopefully later today.

> I've already refactered libbpf-cargo to use this module instead. What do you think is the best way to publish those changes? Should I make a PR that points at this one?

Yeah, feel free to create a PR that is based on this one.

**Review by danielocfb — APPROVED** [review]
I think it looks fine now, thanks again! For the next time, please remember to write proper commit messages.

**Comment by danielocfb** [comment]
I went ahead and merged this patch set (with a few minor changes) -- it will just get even more unwieldy if we built more stuff on top here while the code hasn't landed at an in general it's no fun to update such a big patch set as other changes get merged (although here there is likely sufficient isolation to make this not much of an issue).

**Comment by mendess** [comment]
@danielocfb Thanks for merging!

> I think it looks fine now, thanks again! For the next time, please remember to write proper commit messages.

What could I improve about my commit messages? Hopefully I can do a better job in my next PR.

**Comment by danielocfb** [comment]
> @danielocfb Thanks for merging!
> 
> > I think it looks fine now, thanks again! For the next time, please remember to write proper commit messages.
> 
> What could I improve about my commit messages? Hopefully I can do a better job in my next PR.

The current set of messages only has a title, but no detailed description. But a commit message should really go into more detail why something is done, potentially explain why something is introduced now if it's used only later, provide other context etc. It's mentioned in our contribution guidelines as:

> Include useful details in the commit messages. Providing context helps reviewers understand the change and helps expedite feedback. Please provide a detailed enough description of (high-level) what the change does and why it is needed.

There is no one-size-fits all, of course, but having proper commit messages helps future developers understand changes easier -- which becomes more important as a project matures over time.

We are not the Linux Kernel here (obviously), but if you check its commit history, I am not sure you'll find a single subject-only commit message. There is pretty much always more to be said than just a single sentence. (for reference, here is what its guidelines say: https://www.kernel.org/doc/html/latest/process/submitting-patches.html#describe-your-changes)

**Comment by danielocfb** [comment]
> I'm currently trying to make the examples and tests pass. I think we can consider `libbpf-cargo` pretty good integration testing from what I've seen, it exercises a lot of the API.

@mendess are you still working on that? Just to clarify, me merging this change wasn't meant to convey that testing is no longer necessary and that the overall work is done, but more of a "we are half way there and I trust you to get the rest over the finish line as well, based on what I have seen so far", in effect reducing the amount of churn for subsequent work (and, hence, hopefully making both our lives easier).

**Comment by mendess** [comment]
Yes. I'm still on it. I've just been busy in the mean time. But I have a branch that pretty much only needs commit cleanup

File: libbpf-rs/src/btf/mod.rs (hunk 7 above)
[the remark this answers is not in the document]
Comment (danielocfb) [reply libbpf-rs/src/btf/mod.rs:140]: Sorry, I just came across this again. Doesn't `CString` have a constructor that creates an object from a `Vec`? https://doc.rust-lang.org/std/ffi/struct.CString.html#method.new

I would think something like this would work:
```rust
            CString::new(path.as_ref().as_os_str().as_bytes().to_vec())
```
?
This explicit NUL byte push is really putting me off right now and I see it even being conditional elsewhere...

[the remark this answers is not in the document]
Comment (mendess) [reply libbpf-rs/src/btf/mod.rs:140]: yeah, I did totally miss this obvious route.... we should change this to that version you just posted

[the remark this answers is not in the document]
Comment (mendess) [reply libbpf-rs/src/btf/mod.rs:140]: Should I make a PR or do you want to just push that small change to master directly?

[the remark this answers is not in the document]
Comment (danielocfb) [reply libbpf-rs/src/btf/mod.rs:140]: Feel free to go ahead and create a PR. Will prioritize reviewing it.
